{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7d1692b9-cc23-4b16-b0d5-884ff4f99456",
   "metadata": {},
   "source": [
    "### read fasta files, score alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fec5cf32-b6d9-478f-bb8a-cddfb5c361a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# ----------------------\n",
    "# Load data into memory.\n",
    "# ----------------------\n",
    "\n",
    "# Assign file path to path. \n",
    "path = 'exampledata/score.example.fna'\n",
    "output_path = 'exampledata/output_fasta.txt'\n",
    "\n",
    "# Create dictionary to store FASTA headers and sequences.\n",
    "fasta_dict = {}\n",
    "\n",
    "# Initiate the variable first_line to True, which will be set to False once the first line is read.\n",
    "first_line = True\n",
    "\n",
    "# Define the set valid_chars containing the permitted characters.\n",
    "valid_chars = ['A', 'C', 'G', 'T', '-']\n",
    "\n",
    "# Open the FASTA file.\n",
    "with open(path, 'r') as f:\n",
    "\n",
    "    # Initiate while loop to read lines one at a time.\n",
    "    while True:\n",
    "\n",
    "        # Strip whitespace characters from the start and end of the line and save the string to the variable\n",
    "        # line.\n",
    "        line = f.readline().strip()\n",
    "\n",
    "        # Check if the line starts with '>', in which case it is a header.\n",
    "        if line.startswith('>'):\n",
    "\n",
    "            # Add the previous header and complete sequence read before it to the dictionary fasta_dict.\n",
    "            if not first_line:\n",
    "                fasta_dict[head] = seq\n",
    "\n",
    "            # If the first line of the file is being read, do not add anything to the dictionary since\n",
    "            # a sequence has not been read yet. Set first_line to False.\n",
    "            else:\n",
    "                first_line = False\n",
    "            \n",
    "            # Assign the line to the variable head and set header_read to True.\n",
    "            head = line\n",
    "            header_read = True\n",
    "\n",
    "        # If the line does not start with '>' it is a sequence.\n",
    "        else:\n",
    "            \n",
    "            # Check that the sequence contains only valid characters.\n",
    "            if sum(line.upper().count(c) for c in valid_chars) != len(line):\n",
    "                raise TypeError(f'Invalid characters detected in sequence \"{head.split()[0][1:]}\". '\n",
    "                                f'Program interrupted, please re-align sequences.')\n",
    "\n",
    "            # Check if the previously read line was a header. If so, assign the current line to the\n",
    "            # variable seq.\n",
    "            if header_read:\n",
    "                seq = line.upper()\n",
    "\n",
    "            # If the previously read line was a sequence, add the current line to the previous line to fix\n",
    "            # sequences split by newline characters.\n",
    "            else:\n",
    "                seq += line.upper()\n",
    "\n",
    "            # Assign the variable header_read to False if the line read did not start with a '>'.\n",
    "            header_read = False\n",
    "        \n",
    "        # Check if there are no more lines to read. Add the final head and seq pair to fasta_dict and break\n",
    "        # the while loop.\n",
    "        if not line:\n",
    "            fasta_dict[head] = seq\n",
    "            break\n",
    "\n",
    "\n",
    "# --------------------------           \n",
    "# Generate alignment scores.\n",
    "# --------------------------\n",
    "\n",
    "# Assign scores to variables.\n",
    "match_score = 1\n",
    "transition = -1 # a<->g or c<->t\n",
    "transversion = -2 # other mutations\n",
    "gap_penalty = -1\n",
    "gap_match = 0 # this script assumes that if two sequences both have a gap in the same position, that\n",
    "              # position of the parallel alignment adds 0 to the alignment score.\n",
    "\n",
    "# Assign an empty list to the variable score_summary.\n",
    "score_summary = ''\n",
    "\n",
    "# Loop through the keys in the dictionary, calculating each possible pairwise score.\n",
    "for i, key1 in enumerate(fasta_dict.keys()):\n",
    "    for j, key2 in enumerate(fasta_dict.keys()):\n",
    "\n",
    "        # Only calculate scores when i < j to ensure the same sequence does not get scored with itself or that\n",
    "        # duplicate scores are reported.\n",
    "        if i < j:\n",
    "\n",
    "            # Assign sequences to variables and normalize case. Ensure that they are the same length, otherwise\n",
    "            # interrupt the program.\n",
    "            seq_a, seq_b = fasta_dict[key1], fasta_dict[key2]\n",
    "            if len(seq_a) != len(seq_b):\n",
    "                raise ValueError('sequences are not of the same length. program interrupted')\n",
    "\n",
    "            # Initialize alignment scoring variables.\n",
    "            identity = 0\n",
    "            gaps = 0\n",
    "            score = 0\n",
    "            alignment_len = len(seq_a)\n",
    "\n",
    "            # Zip the sequences to compare alignment at each position and loop through nucleotide by nucleotide.\n",
    "            for a, b in zip(seq_a, seq_b):\n",
    "\n",
    "                # Check if characters are identical.\n",
    "                if a == b:\n",
    "\n",
    "                    # If they represent matching nucleotides, update score and identity count.\n",
    "                    if '-' not in a:\n",
    "                        score += match_score\n",
    "                        identity += 1\n",
    "                        #print(a, b, match_score)\n",
    "\n",
    "                        # IMPORTANT: matched gaps ('-') are ignored\n",
    "                \n",
    "                # If characters are not identical:\n",
    "                else:\n",
    "\n",
    "                    # Check if the unidentical characters represent a nucleotide transition and update the score.\n",
    "                    if len({a, b} & {'A', 'G'}) == 2 or len({a, b} & {'C', 'T'}) == 2:\n",
    "                        score += transition\n",
    "                        #print(a, b, 'transition')\n",
    "\n",
    "                    # The unidentical characters do not represent a nucleotide transition.\n",
    "                    else:\n",
    "\n",
    "                        # Check if the unidentical characters represent a gap and update the score and gap count.\n",
    "                        if '-' in (a, b):\n",
    "                            score += gap_penalty\n",
    "                            gaps += 1\n",
    "                            #print(a, b, 'gap')\n",
    "\n",
    "                        # The last possible combinatinon of characters represents a transversion. Update the score.\n",
    "                        else:\n",
    "                            score += transversion\n",
    "                            #print(a, b, 'transversion')\n",
    "\n",
    "            # Add the score summary to the string score_summary.\n",
    "            score_summary += (f'{key1[1:]}-{key2[1:]}: '\n",
    "                              f'Identity: {identity}/{alignment_len} ({round(identity/alignment_len*100, 1)}%), '\n",
    "                              f'Gaps: {gaps}/{alignment_len} ({round(gaps/alignment_len*100, 1)}%), '\n",
    "                              f'Score={score}\\n')\n",
    "\n",
    "\n",
    "# ------------------\n",
    "# Write output file.\n",
    "# ------------------\n",
    "with open(output_path, 'w') as f:\n",
    "    f.write(score_summary)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
